pkg/gadgets/trace/capabilities/tracer/bpf/capable.bpf.c

```diff
@@ -129,6 +132,9 @@ int BPF_KPROBE(ig_trace_cap_e, const struct cred *cred, struct user_namespace *t
 
 
 	struct args_t args = {};
+	args.current_userns = (u64) BPF_CORE_READ(task, real_cred, user_ns, ns.inum);
+	args.target_userns = (u64) BPF_CORE_READ(targ_ns, ns.inum);
+	BPF_CORE_READ_INTO(&args.cap_effective, task, real_cred, cap_effective.cap[0]);
```

Comment:
Just wondering why `BPF_CORE_READ` is not used in all cases? is there any special on the last one?

Reply:
The amount of bytes copied is different between BPF_CORE_READ and BPF_CORE_READ_INTO:
* BPF_CORE_READ looks at the type of the source variable
* BPF_CORE_READ_INTO looks at the type of the destination variable

For ns.inum, we use BPF_CORE_READ to read "unsigned int inum"  (size: 4 bytes), cast to u64 (size: 8 bytes). The cast just set the 4 most significant bytes to zero. If we were using BPF_CORE_READ_INTO, we would read 4 bytes of the next field in the struct ns_common (`count`):
```
struct ns_common {
        atomic_long_t stashed;
        const struct proc_ns_operations *ops;
        unsigned int inum;
        refcount_t count;
};
```

For cap_effective, we use BPF_CORE_READ_INTO to read 8 bytes (size of the destination variable "args.cap_effective"). We cannot use BPF_CORE_READ to read `cap_effective` in one instruction because of the way it is defined: there are no 8 bytes field but an array. I couldn't get it to work with BPF_CORE_READ.

```
struct cred {
    ...
    kernel_cap_t    cap_effective;
}
#define _LINUX_CAPABILITY_U32S_3     2
#define _KERNEL_CAPABILITY_U32S    _LINUX_CAPABILITY_U32S_3
typedef struct kernel_cap_struct {
        __u32 cap[_KERNEL_CAPABILITY_U32S];
} kernel_cap_t;
```

Reply:
Ah, I was confused. I thought that we're reason the first element of `cap` in `kernel_cap_struct`, but for some reason the caps are represented as an array of two u32s in the kernel. This also answers my question about the data type above. Thanks!